src/plugins/data/server/search/routes/msearch.ts

```diff
@@ -67,16 +67,23 @@ export function registerMsearchRoute(router: IRouter, deps: SearchRouteDependenc
             })
           ),
         }),
+        query: schema.object({
+          data_source_id: schema.maybe(schema.string()),
```

Comment:
Haven't tested this yet, but I see a potential issue:

Scenario:
- Index pattern foo* saved on datasourceA
  - Matches: foo-1, foo-2, etc.
- Index pattern bar* also saved on datasourceA
  - Matches: bar-1, bar-2, etc.
- bar* indices actually exist on datasourceB

Won't this lead to 404s when querying bar* since we're looking in datasourceA but the indices are in datasourceB?

To me i almost think multisearch and multiple data sources is almost fundamentally conflicting ideas. How do we fetch the data from multiple data sources with a single query. this essentially feels like cross cluster searching. unless im missing something

Reply:
You're right, trying to search for an index pattern in a different data source than where it exists will cause a 404 error.

To solve this, we separate the search requests by data source on the client-side before sending them. The `fetchSoon` function groups the requests based on their data source, creating multiple msearch requests, one for each data source.

This way, each msearch request is sent to the correct data source, avoiding 404 errors for indices that are in a different data source.

The internal msearch API needs to call msearch in different OpenSearch clients. They are also expensive when combined in the same request. That's why we only support one data source per request.